### DISCUSSION TITLE
Clang/LLVM 13 and PE exported symbol limits

### DISCUSSION THREAD

**Discussion by jeremyd2019** [msys2/MINGW-packages#9765] [General]
While preparing the update to mingw-w64-clang (really LLVM) 13.0.0, it failed to link libLLVM-13.dll due to:
```
lld-link: error: too many exported symbols (max 65535)
```

This limit is due to the PE/COFF format using a 16-bit integer as in index into the export tables (IIUC).  There are only a couple of options to get around this:
1. Switch to static linking of libLLVM.  It's been a while, but there was a bug that required this to be done here previously, and I recall it making VERY large packages.  I have a CI run going with static linking and will update with the size difference when it's done.
2. Disable enough things to get the number of exports back down under 65535.  I was able to do this by only enabling targets that I felt might be useful on Windows: AArch64;AMDGPU;ARM;NVPTX;WebAssembly;X86.

Either of these options is probably going to negatively impact somebody, so I thought it would be a good idea to open a discussion to see if there's a consensus on what would be the least-bad option here, and to see if there are any suggestions to adjust the list of targets for option 2.

# Package size comparison
(the net upgrade size is from 12.0.1, and most of that on the dynamically-linked side is from the new package "mlir").  Unfortunately the other MSYSTEMs hit the 6 hour GHA limit while packaging/compressing.

### Statically linked (with all default targets) (built locally)
Packages (13) mingw-w64-x86_64-clang-13.0.0-1 mingw-w64-x86_64-clang-analyzer-13.0.0-1 mingw-w64-x86_64-clang-tools-extra-13.0.0-1 mingw-w64-x86_64-compiler-rt-13.0.0-1 mingw-w64-x86_64-libc++-13.0.0-1 mingw-w64-x86_64-libc++abi-13.0.0-1 mingw-w64-x86_64-libunwind-13.0.0-1 mingw-w64-x86_64-lld-13.0.0-1  mingw-w64-x86_64-lldb-13.0.0-1 mingw-w64-x86_64-llvm-13.0.0-1  mingw-w64-x86_64-mlir-13.0.0-1 mingw-w64-x86_64-openmp-13.0.0-1  mingw-w64-x86_64-polly-13.0.0-1
```
Total Installed Size:  5061.76 MiB
Net Upgrade Size:      3916.98 MiB
```

### Dynamically linked (with fewer targets)
  
  Packages (13) mingw-w64-clang-i686-clang-13.0.0-1  mingw-w64-clang-i686-clang-analyzer-13.0.0-1  mingw-w64-clang-i686-clang-tools-extra-13.0.0-1  mingw-w64-clang-i686-compiler-rt-13.0.0-1  mingw-w64-clang-i686-libc++-13.0.0-1  mingw-w64-clang-i686-libc++abi-13.0.0-1  mingw-w64-clang-i686-libunwind-13.0.0-1  mingw-w64-clang-i686-lld-13.0.0-1  mingw-w64-clang-i686-lldb-13.0.0-1  mingw-w64-clang-i686-llvm-13.0.0-1  mingw-w64-clang-i686-mlir-13.0.0-1  mingw-w64-clang-i686-openmp-13.0.0-1  mingw-w64-clang-i686-polly-13.0.0-1
```  
  Total Installed Size:  944.76 MiB
  Net Upgrade Size:      230.62 MiB
```

### Statically linked (with all default targets)

  Packages (13) mingw-w64-clang-i686-clang-13.0.0-1  mingw-w64-clang-i686-clang-analyzer-13.0.0-1  mingw-w64-clang-i686-clang-tools-extra-13.0.0-1  mingw-w64-clang-i686-compiler-rt-13.0.0-1  mingw-w64-clang-i686-libc++-13.0.0-1  mingw-w64-clang-i686-libc++abi-13.0.0-1  mingw-w64-clang-i686-libunwind-13.0.0-1  mingw-w64-clang-i686-lld-13.0.0-1  mingw-w64-clang-i686-lldb-13.0.0-1  mingw-w64-clang-i686-llvm-13.0.0-1  mingw-w64-clang-i686-mlir-13.0.0-1  mingw-w64-clang-i686-openmp-13.0.0-1  mingw-w64-clang-i686-polly-13.0.0-1
```  
  Total Installed Size:  5466.04 MiB
  Net Upgrade Size:      4751.90 MiB
```

**Comment by jeremyd2019** [comment]
A third option that I thought of, but don't know how practical it would be, would be to split libLLVM into multiple DLLs (`BUILD_SHARED_LIBS`?), and then merge the import libraries into a libLLVM-13.dll.a so the consumer wouldn't have to know the difference.

**Reply by mati865** [reply]
`BUILD_SHARED_LIBS` conflicts with `LLVM_LINK_LLVM_DYLIB` and disables building of static libraries.
There is no such option right now so somebody would have to hack it together.

**Comment by jeremyd2019** [comment]
It looks like @mstorsjo only enables AArch64, arm, and X86 targets in his scripts: https://github.com/mstorsjo/llvm-mingw/blob/378525eb9b72a7deacdde084356926724238e231/build-llvm.sh#L235.  Also looks like he uses DYLIB by default, unless explicitly asked not to.

**Reply by mati865** [reply]
>  Also looks like he uses DYLIB by default, unless explicitly asked not to.

Oh, thats a new thing in his LLVM 13 release.

**Reply by mstorsjo** [reply]
Yep, I took that into use recently, and the size benefits are massive.

If not using dylibs, you might be able to trim down some amount of the binary bloat by using minimal frontend wrappers, kinda like what I use. So instead of having two identical large binaries clang.exe and clang++.exe, you'd have a wrapper clang++.exe which invokes `clang-13.exe --driver-mode=g++` which does the same as naming the binary that way. If proper symlinks were used throughout the whole build+packaging+unpacking process, that'd be a non-issue though.

**Reply by mstorsjo** [reply]
Oh, and also, there's some upstream effort lately to try to make the dylib build usable for MSVC too (by marking interfaces more correctly). Marking them properly (so it doesn't end up exporting needless internal symbols) probably would help a bit towards not hitting the 64k exports limit too.

**Comment by jeremyd2019** [comment]
I was able to expand the number of targets while still coming in under 65535 exports (CLANG32 came in at 64249).  The list I came up with is AArch64;AMDGPU;ARM;AVR;BPF;Mips;MSP430;NVPTX;RISCV;WebAssembly;X86.  It might be better to list what *isn't* present from the default list:
* Hexagon (a DSP in Qualcomm CPUs)
* Lanai (some Google-internal CPU)
* PowerPC (old Macs, currently mostly used in automotive applications according to Wikipedia)
* Sparc (old Sun, now Oracle)
* SystemZ (IBM mainframes)
* Xcore (🤷)

**Reply by jeremyd2019** [reply]
I merged with this target list.  It just so happened a version of mesa with fixes for LLVM 13 was merged in the same window, so that was a good opportunity to avoid an extra rebuild of it.

**Comment by jeremyd2019** [comment]
As an update, I later had to disable MIPS too, as rebuilding llvm 13 with clang/llvm 13 resulted in even *more* exported symbols

**Comment by royqh1979** [comment]
Why add targets other than X86/WebAssembly?

Msys2 don't support other platforms, and it seems that no one will do cross-compile under msys2.

And it make programs that statically link with libclang very big (90M+), because clang_createIndex (in libclang) uses llvm::InitializeAllTargets()， and it depends all compiled llvm targets.

**Reply by royqh1979** [reply]
It seems that [mingw-w64 llvm+clang built by qt](https://download.qt.io/development_releases/prebuilt/libclang/) only have X86 targets.

**Comment by MehdiChinoune** [comment]


**Reply by jeremyd2019** [reply]
I know that Zig expects *all* default targets to be enabled and fails to build if they are not.  That ship has obviously already sailed.  I don't know if anybody would be actively trying to cross-compile from msys2 or not, but I would try not to let my failure of imagination limit someone else's ingenuity :wink:

**Reply by jeremyd2019** [reply]
I *think* I read in one of the LLVM docs that NVPTX and AMDGPU were not applicable for Windows.  At least for me, I would rather have ARM than AMDGPU (but I am a luddite who only has very old hardware)

**Reply by jeremyd2019** [reply]
Hmm, I guess since the last time I checked it is possible to build mesa to use AMDGPU, but maybe not to actually run it on Windows?  https://www.phoronix.com/scan.php?page=news_item&px=Mesa-Build-RADV-On-Windows

**Reply by pal1000** [reply]
This means I'll have to check if AMDGPU is available with MSYS2 the same way I do for MSVC and drop RADV from [MinGW package](https://github.com/pal1000/mesa-dist-win) if necessary. 

I'll post an announcement on [pal1000/mesa-dist-win](https://github.com/pal1000/mesa-dist-win) and see if anyone complains.

**Reply by mati865** [reply]
If it works than we should keep it.

**Comment by MehdiChinoune** [comment]
Bad News!
With 15.0.0-rc2 I have to disable AMDGPU and WebAssembly on CLANG64 and BPF, MSP430 and RISCV on UCRT64.

**Reply by mstorsjo** [reply]
Just FYI, I've tried to take precautionary action on this issue (for my builds with only X86, ARM and AArch64, I'm also getting very close to the limit, so I would suspect that by LLVM 16, the dylib build would be broken for me too) - and am implementing something like hidden visibility for COFF; see https://reviews.llvm.org/D130120 and https://reviews.llvm.org/D130121. With those in place, we'd get down to around 35k exports in libLLVM.dll. But that'd only be for builds with clang.

If we land those on git main soon, I think it might be a bit late and disruptive for backporting into the LLVM 15 release at this point, but you might want to carry them as downstream patches already in 15.x.

**Reply by mati865** [reply]
Yeah, I've been considering backporting them since you published them to reviews.llvm.org 😉

**Reply by MehdiChinoune** [reply]
> Just FYI, I've tried to take precautionary action on this issue (for my builds with only X86, ARM and AArch64, I'm also getting very close to the limit, so I would suspect that by LLVM 16, the dylib build would be broken for me too) - and am implementing something like hidden visibility for COFF; see https://reviews.llvm.org/D130120 and https://reviews.llvm.org/D130121. With those in place, we'd get down to around 35k exports in libLLVM.dll. But that'd only be for builds with clang.
> 
> If we land those on git main soon, I think it might be a bit late and disruptive for backporting into the LLVM 15 release at this point, but you might want to carry them as downstream patches already in 15.x.

How to use those patches, I have applied them but still exceeding symbols limit.
Should I bootstrap it (Build with just one target and then use that new compiler to build one with all targets)?
Should I pass `--exclude-symbols` to `lld`?

**Reply by mati865** [reply]
First you need to build LLVM 14 with all three patches (maybe also https://reviews.llvm.org/D130200 if you want to be still able to rebuild LLVM 14 with itself), then use that to build LLVM 15.
Alternatively you can build LLVM 15 with 3 patches applied and without shared libs, then use it to rebuild LLVM 15 with shared libs.

**Reply by mstorsjo** [reply]
Yep, exactly like this. And then, for the second stage, you need to apply another patch on llvm, to take the hidden visibility into use, that I haven’t posted for review yet.  See https://martin.st/temp/0001-llvm-Use-hidden-visibility-when-building-on-MinGW.patch.

**Comment by MehdiChinoune** [comment]
New bad News!
```
1395/1397] Linking CXX shared library bin\libMLIR.dll
 FAILED: bin/libMLIR.dll tools/mlir-shlib/libMLIR.dll.a 
 cmd.exe /C "cd . && C:\msys64\ucrt64\bin\g++.exe -march=x86-64 -mtune=generic -O2 -pipe -Wa,-mbig-obj -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -Wimplicit-fallthrough -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wmisleading-indentation -ffunction-sections -fdata-sections -O3 -DNDEBUG  -pipe   -Wl,--gc-sections -shared -o bin\libMLIR.dll -Wl,--out-implib,tools\mlir-shlib\libMLIR.dll.a -Wl,--major-image-version,0,--minor-image-version,0 @CMakeFiles\MLIR.rsp  && cd ."
 C:/msys64/ucrt64/lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: error: export ordinal too large: 77799
 
 collect2.exe: error: ld returned 1 exit status
 ninja: build stopped: subcommand failed.
```
```
  [1388/1397] Linking CXX shared library bin\libMLIR.dll
  FAILED: bin/libMLIR.dll tools/mlir-shlib/libMLIR.dll.a 
  cmd.exe /C "cd . && C:\msys64\clang64\bin\clang++.exe -march=x86-64 -mtune=generic -O2 -pipe -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -ffunction-sections -fdata-sections -Werror=mismatched-tags -O3 -DNDEBUG  -pipe   -Wl,--gc-sections -shared -o bin\libMLIR.dll -Wl,--out-implib,tools\mlir-shlib\libMLIR.dll.a -Wl,--major-image-version,0,--minor-image-version,0 @CMakeFiles\MLIR.rsp  && cd ."
  ld.lld: error: too many exported symbols (got 83452, max 65535)
  
  clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```
I don't know what to do with MLIR?

**Reply by mati865** [reply]
As Martin already noticed it's impossible to build MLIR without patches so there are 2 ways: disable or patch LLVM.

**Reply by mstorsjo** [reply]
Also, even with those patches for reducing the number of exports, libMLIR*.dll does link successfully, but other things (I don’t remember if it was in MLIR or flang or both) still fail to build with dylibs enabled, that are trickier to sort out - I’ve got some wip patches for that that I haven’t posted yet either.

**Reply by mstorsjo** [reply]
Oh, and the patches won’t help for the gcc build, as gcc doesn’t handle the hidden visibility.

**Reply by mati865** [reply]
"Mitigations" for LLVM are in place https://github.com/msys2/MINGW-packages/pull/12536 but I'm wondering what to do with GCC based builds.
Some of the options are:
- disable problematic targets/projects
- backport Binutils patch and build with Clang+ld.bfd
- build with Clang+LLD and ship import libraries that will cause issues
- build with Clang+LLD and recreate all import libs with dlltool
- build with Clang+LLD and delete import libraries (this would rely on direct DLL linking)

**Reply by mstorsjo** [reply]
I guess disabling dylibs also is one option to add to that list, but it's certainly not any better than the other ones.

**Reply by mstorsjo** [reply]
Out of those options, I think backporting the binutils patch and building with Clang+ld.bfd is the most straightforward solution. (And if you do that, do annotate the hack with a comment saying that it is due to the dylibs build, and can be removed if GCC gets support for the same solution.)

**Reply by MehdiChinoune** [reply]
> * build with Clang+LLD and ship import libraries that will cause issues

What kind of issues?

**Reply by mstorsjo** [reply]
> > * build with Clang+LLD and ship import libraries that will cause issues
> 
> What kind of issues?

LLD produces import libraries in the MSVC format. GNU binutils do handle linking against such import libraries, but they don't work perfectly in all cases. (At least using the import libraries for libmsvcrt* from a llvm based sysroot doesn't work with binutils. For linking against a regular user DLL like libLLVM, it might still work, I'm not sure.)

**Reply by mati865** [reply]
> Out of those options, I think backporting the binutils patch and building with Clang+ld.bfd is the most straightforward solution. (And if you do that, do annotate the hack with a comment saying that it is due to the dylibs build, and can be removed if GCC gets support for the same solution.)

Yeah, I have proposed this route on MSYS2 discord in the morning.



> > > * build with Clang+LLD and ship import libraries that will cause issues
> > 
> > 
> > What kind of issues?
> 
> LLD produces import libraries in the MSVC format. GNU binutils do handle linking against such import libraries, but they don't work perfectly in all cases. (At least using the import libraries for libmsvcrt* from a llvm based sysroot doesn't work with binutils. For linking against a regular user DLL like libLLVM, it might still work, I'm not sure.)

I have seen too many times where it works with single short import lib (like one made by LLVM) but adding second short import lib produces executable with mangled import table.

**Reply by mstorsjo** [reply]
> > > > * build with Clang+LLD and ship import libraries that will cause issues
> > > 
> > > 
> > > What kind of issues?
> > 
> > 
> > LLD produces import libraries in the MSVC format. GNU binutils do handle linking against such import libraries, but they don't work perfectly in all cases. (At least using the import libraries for libmsvcrt* from a llvm based sysroot doesn't work with binutils. For linking against a regular user DLL like libLLVM, it might still work, I'm not sure.)
> 
> I have seen too many times where it works with single short import lib (like one made by LLVM) but adding second short import lib produces executable with mangled import table.

FWIW I also has such faint memories, but when I've tried to dig in closer in such issues, I only found a temporary regression that caused such a situation in binutils 2.34, but with current versions it seems fine, at least with small minimal examples. If you ever feel like trying such stuff again, please do, and report any such errors so that someone can try to narrow it down. I'd love to get better interoperability between these two formats.

**Reply by lazka** [reply]
(the binutils patches are backported now: https://github.com/msys2/MINGW-packages/pull/12556)

**Reply by mati865** [reply]
@mstorsjo I have extracted 2 Binutils failures from Rust's testsuite:
There are 2 tests:
- [raw-dylib-c](https://github.com/rust-lang/rust/tree/0a754b309ce05a34a10aac339c1a4253b7a551ee/src/test/run-make/raw-dylib-c)
- [raw-dylib-link-ordinal](https://github.com/rust-lang/rust/tree/0a754b309ce05a34a10aac339c1a4253b7a551ee/src/test/run-make/raw-dylib-link-ordinal)

In attached ZIP there are binaries (sorry, couldn't quickly get reproducer that doesn't involve patched Rustc) in directories with name:
- `dlltool-imp-lib` - import library was made with Binutils `dlltool` from `.def` file and linked by `ld.bfd`
- `short-imp-lib` - import library was made with LLVM directly and linked with `ld.bfd`
- `short-imp-lib-lld` - import library was made with LLVM directly and linked with `ld.lld`

`dlltool-imp-lib` and `short-imp-lib-lld` seem to work flawlessly but `short-imp-lib` crash:

<details>
<summary>Ran under LLDB 15 RC2:</summary>

```
(lldb) target create "build/x86_64-pc-windows-gnu/test/run-make/raw-dylib-c/raw-dylib-c/driver.exe"
Current executable set to 'D:\Projekty\rust\build\x86_64-pc-windows-gnu\test\run-make\raw-dylib-c\raw-dylib-c\driver.exe' (x86_64).
(lldb) r
(lldb) Process 30636 launched: 'D:\Projekty\rust\build\x86_64-pc-windows-gnu\test\run-make\raw-dylib-c\raw-dylib-c\driver.exe' (x86_64)
(lldb) Process 30636 stopped
* thread #1, stop reason = Exception 0xc0000005 encountered at address 0x0fa7d8: User-mode data execution prevention (DEP) violation at location 0x000fa7d8
    frame #0: 0x00000000000fa7d8
error: Only part of a ReadProcessMemory or WriteProcessMemory request was completed.
(lldb) bt
error: driver.exe {0x000022af}: DIE has DW_AT_ranges(DW_FORM_sec_offset 0x1a) attribute, but range extraction failed (invalid range list offset 0x1a), please file a bug and attach the file at the start of this error message
(thread #1, stop reason = Exception 0xc0000005 encountered at address 0x0fa7d8: User-mode data execution prevention (DEP) violation at location 0x000fa7d8
  * frame #0: 0x00000000000fa7d8
    frame #1: 0x00007ff7b68c16d9 driver.exe`raw_dylib_test::library_function::h4a674cc82b6bf4d0 + 9
    frame #2: 0x00007ff7b68c1629 driver.exe`driver::main::heb9640d0c9cabce8 + 9
    frame #3: 0x00007ff7b68c15a6 driver.exe`core::ops::function::FnOnce::call_once::he991f2c0d61fc9d4 + 6
    frame #4: 0x00007ff7b68c1619 driver.exe`std::sys_common::backtrace::__rust_begin_short_backtrace::h2f2cf564d2cb4782 + 9
    frame #5: 0x00007ff7b68c15fc driver.exe`std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::hb23e9c2a34583301 + 12
    frame #6: 0x00007ff7b68da88e driver.exe`std::rt::lang_start_internal::h3e01f0df7200482a + 30
    frame #7: 0x00007ff7b68c15e0 driver.exe`std::rt::lang_start::h916567bfeec30977 + 32
    frame #8: 0x00007ff7b68c165c driver.exe`main + 44
    frame #9: 0x00007ff7b68c13ae driver.exe`__tmainCRTStartup
    frame #10: 0x00007ff7b68c14e6 driver.exe`mainCRTStartup
    frame #11: 0x00007ffc7d3254e0 kernel32.dll`BaseThreadInitThunk + 16
    frame #12: 0x00007ffc7d7e485b ntdll.dll`RtlUserThreadStart + 43
error: driver.exe {0x000022af}: DIE has DW_AT_ranges(DW_FORM_sec_offset 0x1a) attribute, but range extraction failed (invalid range list offset 0x1a), please file a bug and attach the file at the start of this error message
(lldb)(lldb)  d
error: Only part of a ReadProcessMemory or WriteProcessMemory request was completed.
error: Failed to disassemble memory at 0x000fa7d8.
```

```
(lldb) target create "build/x86_64-pc-windows-gnu/test/run-make/raw-dylib-link-ordinal/raw-dylib-link-ordinal/driver.exe"
(rrent executable set to 'D:\Projekty\rust\build\x86_64-pc-windows-gnu\test\run-make\raw-dylib-link-ordinal\raw-dylib-link-ordinal\driver.exe' (x86_64).
(lldb) r
(lldb) Process 36388 launched: 'D:\Projekty\rust\build\x86_64-pc-windows-gnu\test\run-make\raw-dylib-link-ordinal\raw-dylib-link-ordinal\driver.exe' (x86_64)
Process 36388 stopped
* thread #1, stop reason = Exception 0xc0000005 encountered at address 0x7ff7c30d1704: Access violation reading location 0xffffffffffffffff
    frame #0: 0x00007ff7c30d1704 driver.exe
->  0x7ff7c30d1704: jmpq   *0xf84ce(%rip)            ; .idata$5
    0x7ff7c30d170a: nop
    0x7ff7c30d170b: nop
    0x7ff7c30d170c: jmpq   *0xf84ce(%rip)            ; .idata$5
(lldb) bt
error: driver.exe {0x000022af}: DIE has DW_AT_ranges(DW_FORM_sec_offset 0x1a) attribute, but range extraction failed (invalid range list offset 0x1a), please file a bug and attach the file at the start of this error message
(thread #1, stop reason = Exception 0xc0000005 encountered at address 0x7ff7c30d1704: Access violation reading location 0xffffffffffffffff
  * frame #0: 0x00007ff7c30d1704 driver.exe
    frame #1: 0x00007ff7c30d13ae driver.exe`__tmainCRTStartup
    frame #2: 0x00007ff7c30d14e6 driver.exe`mainCRTStartup
    frame #3: 0x00007ffc7d3254e0 kernel32.dll`BaseThreadInitThunk + 16
    frame #4: 0x00007ffc7d7e485b ntdll.dll`RtlUserThreadStart + 43
(lldb)error: driver.exe {0x000022af}: DIE has DW_AT_ranges(DW_FORM_sec_offset 0x1a) attribute, but range ext raction failed (invalid range list offset 0x1a), please file a bug and attach the file at the start of this error message
(lldb) d
->  0x7ff7c30d1704: jmpq   *0xf84ce(%rip)            ; .idata$5
    0x7ff7c30d170a: nop
    0x7ff7c30d170b: nop
    0x7ff7c30d170c: jmpq   *0xf84ce(%rip)            ; .idata$5
    0x7ff7c30d1712: nop
    0x7ff7c30d1713: nop
    0x7ff7c30d1714: jmpq   *0xf84ce(%rip)            ; .idata$5
    0x7ff7c30d171a: nop
    0x7ff7c30d171b: nop
    0x7ff7c30d171c: nopl   (%rax)
```

</details>

[raw-dylib-tests.zip](https://github.com/msys2/MINGW-packages/files/9378047/raw-dylib-tests.zip)

I can try to get C MCVE soon if those are not helpful.

**Reply by mstorsjo** [reply]
@mati865 Thanks for those testcases. When doing e.g. `llvm-readobj --coff-imports raw-dylib-c-short-imp-lib/driver.exe`, I see e.g. this, which indeed seems messed up:
```
Import {
  Name: extern_1.dll
  ImportLookupTableRVA: 0xF9668
  ImportAddressTableRVA: 0xF9C20
}
Import {
  Name: extern_2.dll
  ImportLookupTableRVA: 0xF9678
  ImportAddressTableRVA: 0xF9C30
}
```
(No actual symbols imported from each of them.)

To actually reproduce and fix the issue, I would need to be able to relink the test executable, i.e. have all the input object files/libraries needed for it (modulo standard mingw import libraries where I presume any copy would work the same). Are these libraries in this case linked against `libraw_dylib_test.rlib`, which is a combined import library containing the import libraries for both extern_1.dll and extern_2.dll? I guess it's plausible that that would be problematic (and binutils would need to support that for handling linking against e.g. libucrt.a, which is a combination of import libraries for api-ms-win-crt-*.dll too.

I guess it should be straightforward to construct such a testcase from scratch - let me try...

**Reply by mstorsjo** [reply]
Yes, I can reproduce that situation by merging two short import libraries into one and linking it with binutils ld.bfd.

**Reply by mati865** [reply]
AFAIK `libraw_dylib_test.rlib` is just a static library with Rust symbols.
Executables have been created with temporary import library (which is not included because compiler has already removed it). I'll try to get it later today.

**Reply by mstorsjo** [reply]
No, `libraw_dylib_test.rlib` seems to be a combination, containing statically linked rust code, and also bundled import libraries; if you list the members of the archive with e.g. `ar t libraw_dylib_test.rlib`, you'll find two members named `lib.rmeta` and `raw_dylib_test.raw_dylib_test.649a6ce8-cgu.0.rcgu.o`, followed by regular (long or short) import library members for both libraries.

**Reply by mati865** [reply]
Ah, you are right. I've confused this test with another one that probably no longer exists.

**Reply by mati865** [reply]
Tested it further and even combining single short import library with static library results in the same messed up import table.

**Reply by mstorsjo** [reply]
Actually, a short import library with some static object files should have been supported - as long as those object files were named `*.obj` - but if not, you'd hit the same issue.

I finally got my head wrapped around this logic, and it seems possible to simplify it a lot, getting rid of the requirement of the static object files being named `*.obj` at the same time as fixing support for more than one DLL in the same import library. Can you testrun my patches for me (to see if they generally seem to work)? It's the two topmost commits at https://github.com/mstorsjo/binutils-gdb/commits/msvc-implib3.

**Reply by mati865** [reply]
It failed with:
```
ei386pe.c:153:12: error: static declaration of 'stricmp' follows non-static declaration
  153 | static int stricmp (const char *s1, const char *s2)
      |            ^~~~~~~
In file included from D:/msys64/mingw64/include/io.h:10,
                 from D:/msys64/mingw64/include/sys/stat.h:14,
                 from ../../binutils-2.39/ld/sysdep.h:35,
                 from ei386pe.c:35:
D:/msys64/mingw64/include/string.h:110:15: note: previous declaration of 'stricmp' with type 'int(const char *, const char *)'
  110 |   int __cdecl stricmp(const char *_Str1,const char *_Str2) __MINGW_ATTRIB_DEPRECATED_MSVC2005;
      |               ^~~~~~~
```

Then I renamed the functions and all calls within these files to `__stricmp` and it passed Rust tests with import libraries created with LLVM! This is wonderful.

I'm running full bootstrap with more tests to catch potential regressions so if I don't reply within an hour that will mean there were no regressions (from Rust POV at least).

**Reply by mstorsjo** [reply]
As you didn't follow up, I presume this worked well in the end! And I guess it works fine for other regular libraries too (gnu style import libraries and normal static libraries) as your build would have contained a bunch of them too. So in that case, I think I'll try to send this patch upstream (the first one about case insensitivity got approved already).

**Reply by mati865** [reply]
Yeah, 2 stage bootstrap + testsuite found no regressions.

> And I guess it works fine for other regular libraries too (gnu style import libraries and normal static libraries) as your build would have contained a bunch of them too

Regular libraries were overwhelming majority in fact.

**Reply by mstorsjo** [reply]
FYI, this patch was approved and merged upstream now.
